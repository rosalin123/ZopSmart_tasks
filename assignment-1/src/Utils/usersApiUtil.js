const baseUrl = 'https://jsonplaceholder.typicode.com';

export const fetchUsers = async () => {
  try {
    let url = `${baseUrl}/users`;
    let users = await fetch(url, { method: 'GET' });
    let usersJson = await users.json();
    return usersJson;
  } catch (err) {
    console.log('Error fetching users details', err.stack);
  }
};

export const getUser = async (id) => {

    let url = `${baseUrl}/users/${id}`;
    let user = await fetch(url, { method: 'GET' });
    let userJson = await user.json();
    return userJson;

};
