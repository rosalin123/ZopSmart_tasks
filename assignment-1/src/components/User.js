import React, { Component } from 'react';
import { getUser } from '../actions/userActions';
import { fetchUserPosts } from '../actions/postActions';
import { connect } from 'react-redux';
import { withStyles } from '@material-ui/core/styles';
import Avatar from '@material-ui/core/Avatar';
import { Grid, Typography } from '@material-ui/core';
import PostComponent from './PostComponent';
import { deepOrange } from '@material-ui/core/colors';
import Button from '@material-ui/core/Button';
import Loader from './loader';
import ErrorComponent from './error';
import About from './about.js';

const styles = {
  root: {
    width: '100%',
  },

  square: {
    position: 'relative',
    backgroundImage: 'linear-gradient(to bottom,black, lightgrey)',
    width: '100vw',
    height: 500,
  },

  lineStyles: {
    width: 800,
  },
  avatar: {
    position: 'absolute',
    bottom: -100,
    left: '43.75%',
    width: 200,
    height: 200,
    border: '10px solid white',
    backgroundColor: deepOrange[500],
    fontSize: '70px',
    fontWeight: 'bold',
  },

  userNameStyles: {
    paddingTop: 120,
    paddingBottom: 20,
    width: '100%',
    background: 'white',
    textAlign: 'center',
  },

  postStyles: {
    background: '#eceff1',
    width: '100vw',
  },

  aboutStyles: {
    background: '#eceff1',
    marginTop: '-16px',
    padding: '16px',
    width: '100vw',
  },

  typographyStyles: {
    fontWeight: 'bold',
    fontSize: '28px',
    textAlign: 'center',
  },
};

class User extends Component {
  state = {
    showPosts: true,
    showAbout: false,
  };

  componentDidMount = () => {
    let { id } = this.props.match.params;
    this.props.getUser(id);
    this.props.fetchUserPosts(id);
  };

  handleClick = (e) => {
    if (e.currentTarget.textContent === 'About') {
      this.setState({ showAbout: true, showPosts: false });
    } else {
      this.setState({ showAbout: false, showPosts: true });
    }
  };

  render() {
    const { classes } = this.props;

    let {
      posts,
      user,
      userErr,
      postsErr,
      userLoading,
      postsLoading,
    } = this.props;
    return userLoading === true || postsLoading === true ? (
      <Loader />
    ) : userErr !== '' || postsErr !== '' ? (
      <ErrorComponent message="Something's not right" />
    ) : (
      <Grid container>
        <Grid
          container
          direction="column"
          justify="center"
          alignItems="center"
          className={classes.root}
        >
          <Grid item>
            {' '}
            <div className={classes.square}>
              <Avatar className={classes.avatar}>
                {user.username ? user.username[0] : ''}{' '}
              </Avatar>
            </div>
          </Grid>
          <Grid item className={classes.userNameStyles}>
            <Typography className={classes.typographyStyles}>
              {user.username}
            </Typography>

            <Typography color="textSecondary">{user.email}</Typography>

            <hr className={classes.lineStyles} />

            <Grid container justify="center" spacing={5}>
              <Grid item>
                <Button color="inherit" onClick={(e) => this.handleClick(e)}>
                  Posts
                </Button>
              </Grid>
              <Grid item>
                <Button color="inherit" onClick={(e) => this.handleClick(e)}>
                  About
                </Button>
              </Grid>
            </Grid>
          </Grid>

          {this.state.showPosts === true ? (
            <Grid item>
              <Grid
                container
                spacing={4}
                justify="center"
                alignItems="center"
                className={classes.postStyles}
                direction="column"
              >
                {posts.map((post, index) => {
                  return (
                    <Grid item key={index}>
                      <PostComponent
                        title={post.title}
                        user={user}
                        body={post.body}
                        id={post.id}
                      />
                    </Grid>
                  );
                })}
              </Grid>
            </Grid>
          ) : (
            <Grid item className={classes.aboutStyles}>
              <Grid container justify="center">
                <Grid item>
                  <About user={user} />
                </Grid>
              </Grid>
            </Grid>
          )}
        </Grid>
      </Grid>
    );
  }
}

const mapStateToProps = (state) => ({
  user: state.User.user,
  userLoading: state.User.loading,
  userErr: state.User.error,
  posts: state.UserPosts.posts,
  postsLoading: state.UserPosts.loading,
  postsErr: state.UserPosts.error,
});

const mapDispatchToProps = (dispatch) => {
  return {
    getUser: (userId) => dispatch(getUser(userId)),
    fetchUserPosts: (userId) => dispatch(fetchUserPosts(userId)),
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withStyles(styles)(User));
