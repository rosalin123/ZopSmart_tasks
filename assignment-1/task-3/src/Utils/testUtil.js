import React from 'react';

export const findbyTestAttr = (component, attr) => {
  const wrapper = component.find(`[data-test='${attr}']`);
  return wrapper;
};
