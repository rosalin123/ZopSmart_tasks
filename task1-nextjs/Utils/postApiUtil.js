import axios from 'axios';
import fetch from 'isomorphic-unfetch';
const baseUrl = 'https://jsonplaceholder.typicode.com';

export const fetchPosts = async () => {
  let url = `${baseUrl}/posts`;
  let posts = await axios.get(url);
  let postsJson = await posts.json();
  return postsJson;
};

export const getPost = async (id) => {
  let url = `${baseUrl}/posts/${id}`;
  let post = fetch(url);
  let postJson = await post.json();
  return postJson;
};

export const getPostComments = async (id) => {
  let url = `${baseUrl}/posts/${id}/comments`;
  let comments = await axios.get(url);
  let commentsJson = await comments.json();
  return commentsJson;
};

export const getUserPosts = async (id) => {
  let url = `${baseUrl}/posts?userId=${id}`;
  let userPosts = await fetch(url);
  let userPostsJson = await userPosts.json();
  return userPostsJson;
};
