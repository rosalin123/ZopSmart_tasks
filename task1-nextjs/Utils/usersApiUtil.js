import fetch from 'isomorphic-unfetch';
const baseUrl = 'https://jsonplaceholder.typicode.com';

export const fetchUsers = async () => {
  try {
    let url = `${baseUrl}/users`;
    let users = await fetch(url);
    let usersJson = await users.json();
    return usersJson;
  } catch (err) {
    console.log('Error fetching users details', err.stack);
  }
};

export const getUser = async (id) => {
  let url = `${baseUrl}/users/${id}`;
  let user = await fetch(url);
  let userJson = await user.json();
  return userJson;
};
