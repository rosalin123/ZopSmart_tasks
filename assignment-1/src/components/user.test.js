import React from 'react';
import { render, shallow, mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import User, { User as UserComponent } from './User';
import { BrowserRouter as Router } from 'react-router-dom';
import { findbyTestAttr } from '../Utils/testUtil';
import { mapDispatchToProps } from './User';
import { middleware } from '../store';

const mockStore = configureMockStore(middleware);

const setUp = (initialState) => {
  const store = mockStore(initialState);
  const component = render(
    <Router>
      <Provider store={store}>
        <User />
      </Provider>
    </Router>
  );
  return component;
};

describe('User Component', () => {
  let component;

  describe('renders user component when when loading is false and no errors have occured  ', () => {
    const initialState = {
      User: {
        loading: false,
        user: { name: 'ros', username: 'ros', email: 'r@gmail.com' },
        error: '',
      },
      UserPosts: {
        loading: false,
        posts: [{ title: 'title 1', body: 'post 1' }],
        error: '',
      },
    };

    beforeEach(() => {
      component = setUp(initialState);
    });

    it('should render user component without any error', () => {
      expect(component).toMatchSnapshot();
    });

    it('should render username', () => {
      const wrapper = findbyTestAttr(component, 'username');
      expect(wrapper.length).toBe(1);
      expect(wrapper.text()).toEqual('ros');
    });

    it('should render navbar', () => {
      expect(findbyTestAttr(component, 'navbar').length).toBe(1);
    });

    it('should render avatar', () => {
      expect(findbyTestAttr(component, 'avatar').length).toBe(1);
    });

    it('should render email', () => {
      const wrapper = findbyTestAttr(component, 'useremail');
      expect(wrapper.length).toBe(1);
      expect(wrapper.text()).toEqual('r@gmail.com');
    });
  });

  describe('component did mount and handleClick method', () => {
    const props = {
      match: { params: { id: 1 } },
      getUser: jest.fn(),
      fetchUserPosts: jest.fn(),
      user: { name: 'ros', username: 'ros', email: 'r@gmail.com' },
      userLoading: false,
      userErr: '',
      posts: [{ title: 'title 1', body: 'post 1' }],
      postsLoading: false,
      postsErr: '',
      classes: {
        root: {},
        square: {},
        lineStyles: {},
        avatar: {},
        userNameStyles: {},
        postStyles: {},
        aboutStyles: {},
        typographyStyles: {},
      },
    };

    beforeEach(() => {
      component = shallow(<UserComponent {...props} />);
    });

    it('handleClick should work as expected', () => {
      let postButton = findbyTestAttr(component, 'postsButton');
      let aboutButton = findbyTestAttr(component, 'aboutUserButton');
      expect(postButton.length).toBe(1);
      expect(aboutButton.length).toBe(1);
      postButton.simulate('click');
      expect(component.state('showPosts')).toBe(true);
      expect(component.state('showAbout')).toBe(false);
      aboutButton.simulate('click');
      expect(component.state('showPosts')).toBe(false);
      expect(component.state('showAbout')).toBe(true);
    });

    it('componentDidMount should call the methods as expected', () => {
      const instance = component.instance();

      jest.spyOn(instance.props, 'getUser');
      jest.spyOn(instance.props, 'fetchUserPosts');
      instance.componentDidMount();
      expect(instance.props.getUser).toHaveBeenCalledTimes(1);
      expect(instance.props.fetchUserPosts).toHaveBeenCalledTimes(1);
    });
  });

  it(' should render loader when loading is true', () => {
    const initialState = {
      User: {
        loading: true,
        error: '',
      },
      UserPosts: {
        loading: false,
        error: '',
      },
    };
    component = setUp(initialState);
    expect(findbyTestAttr(component, 'loader').length).toBe(1);
  });

  it('should render ErrorComponent when error occurs', () => {
    const initialState = {
      User: {
        loading: false,
        error: 'Something wrong',
      },
      UserPosts: {
        loading: false,
        error: '',
      },
    };
    component = setUp(initialState);

    let wrapper = findbyTestAttr(component, 'errorComponent');
    expect(wrapper.length).toBe(1);
    expect(findbyTestAttr(component, 'loader').length).toBe(0);
  });

  it('mapDispatchToProps', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).getUser(1);
    mapDispatchToProps(dispatch).fetchUserPosts(1);
    expect(typeof dispatch.mock.calls[0][0]).toBe('function');
    expect(typeof dispatch.mock.calls[1][0]).toBe('function');
  });
});
