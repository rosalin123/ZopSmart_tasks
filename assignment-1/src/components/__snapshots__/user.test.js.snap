// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Component renders user component when when loading is false and no errors have occured   should render user component without any error 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container"
    data-test="userComponent"
  >
    <div
      class="MuiGrid-root User-root-1 MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
    >
      <div
        class="MuiGrid-root MuiGrid-item"
      >
         
        <div
          class="User-square-2"
          data-test="avatar"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circle User-avatar-4 MuiAvatar-colorDefault"
          >
            r 
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root User-userNameStyles-5 MuiGrid-item"
      >
        <p
          class="MuiTypography-root User-typographyStyles-8 MuiTypography-body1"
          data-test="username"
        >
          ros
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          data-test="useremail"
        >
          r@gmail.com
        </p>
        <hr
          class="User-lineStyles-3"
        />
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-5 MuiGrid-justify-xs-center"
          data-test="navbar"
        >
          <div
            class="MuiGrid-root MuiGrid-item"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
              data-test="postsButton"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Posts
              </span>
            </button>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
              data-test="aboutUserButton"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                About
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item"
        data-test="posts"
      >
        <div
          class="MuiGrid-root User-postStyles-6 MuiGrid-container MuiGrid-spacing-xs-4 MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
        >
          <div
            class="MuiGrid-root MuiGrid-item"
          >
            <div
              class="MuiPaper-root MuiCard-root makeStyles-root-9 MuiPaper-elevation1 MuiPaper-rounded"
              data-test="postDetails"
            >
              <div
                class="MuiCardHeader-root"
                data-test="postHeader"
              >
                <div
                  class="MuiCardHeader-avatar"
                >
                  <div
                    aria-label="recipe"
                    class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-15 MuiAvatar-colorDefault"
                  >
                    r
                  </div>
                </div>
                <div
                  class="MuiCardHeader-content"
                >
                  <span
                    class="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
                  >
                    <p
                      class="makeStyles-cardHeaderStyles-11"
                      style="margin:0;padding:0"
                    >
                      ros
                    </p>
                  </span>
                  <span
                    class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                  >
                    September 14, 2016
                  </span>
                </div>
                <div
                  class="MuiCardHeader-action"
                >
                  <button
                    aria-label="settings"
                    class="MuiButtonBase-root MuiIconButton-root"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                        />
                      </svg>
                    </span>
                  </button>
                </div>
              </div>
              <div
                class="MuiCardContent-root"
                data-test="postBody"
              >
                 
                <a
                  class="makeStyles-linkStyles-16"
                  data-test="postLink"
                  href="/posts/undefined"
                >
                  <p
                    class="MuiTypography-root makeStyles-cardContentTitleStyles-12 MuiTypography-body2 MuiTypography-colorTextPrimary"
                    data-test="postTitle"
                  >
                    title 1
                  </p>
                </a>
                <p
                  class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
                  data-test="postContent"
                >
                  post 1
                </p>
              </div>
              <div
                class="MuiCardActions-root"
              >
                <button
                  aria-label="add to favorites"
                  class="MuiButtonBase-root MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
                      />
                    </svg>
                  </span>
                </button>
                <button
                  aria-label="share"
                  class="MuiButtonBase-root MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81 1.66 0 3-1.34 3-3s-1.34-3-3-3-3 1.34-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9c-1.66 0-3 1.34-3 3s1.34 3 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.16c-.05.21-.08.43-.08.65 0 1.61 1.31 2.92 2.92 2.92 1.61 0 2.92-1.31 2.92-2.92s-1.31-2.92-2.92-2.92z"
                      />
                    </svg>
                  </span>
                </button>
                <button
                  aria-expanded="false"
                  aria-label="show more"
                  class="MuiButtonBase-root MuiIconButton-root makeStyles-expand-13"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                      />
                    </svg>
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
   
</div>
`;
